### DISCUSSION TITLE
How can I get TS typing happy with a specified `id` for child machine?

### DISCUSSION THREAD

**Discussion by martynchamberlin** [statelyai/xstate#3825] [Q&A] [answered]
Using the "minute machine / parent machine" example at https://xstate.js.org/docs/guides/communication.html#invoking-machines, I am trying to figure out how to specify an `id` in a way that gets TS happy. I have `minute-machine.ts`: 

```ts
/* minute-machine.ts */
import { createMachine } from 'xstate';

export const minuteMachine = createMachine({
  id: 'timer',
  // eslint-disable-next-line @typescript-eslint/consistent-type-imports
  tsTypes: {} as import('./minute-machine.typegen').Typegen0,
  initial: 'active',
  states: {
    active: {
      after: {
        60000: { target: 'finished' },
      },
    },
    finished: { type: 'final' },
  },
});
```

Then I'm importing it into `parent-machine.ts` like so: 

```ts
/* parent-machine.ts */
import { createMachine, interpret } from 'xstate';
import { minuteMachine } from './minute-machine';

export const parentMachine = createMachine({
  id: 'parent',
  // eslint-disable-next-line @typescript-eslint/consistent-type-imports
  tsTypes: {} as import('./parent-machine.typegen').Typegen0,
  initial: 'pending',
  states: {
    pending: {
      invoke: {
        id: 'minuteMachine',
        src: minuteMachine,
        // The onDone transition will be taken when the
        // minuteMachine has reached its top-level final state.
        onDone: 'timesUp',
      },
    },
    timesUp: {
      type: 'final',
    },
  },
});

export const service = interpret(parentMachine);
```

On that last line, I'm getting this error: 

> Argument of type 'StateMachine<unknown, any, AnyEventObject, { value: any; context: unknown; }, BaseActionObject, ServiceMap, ResolveTypegenMeta<Typegen0, AnyEventObject, BaseActionObject, ServiceMap>>' is not assignable to parameter of type '"Some implementations missing"'.ts(2345)

If I remove the `id: 'minuteMachine',` then the problem goes away. But I need that ID, so I can directly `send()` events to the child machine. Any ideas on what I need to do? I assume it's something I'd pass in `schema` but I cannot figure out what that might need to be. Not seeing any examples in the docs on this. Thanks!

**Comment by Andarist** [comment, marked as answer]
It looks like a bug in our typegen - but I would love to see a full repro case (a runnable one, for example in a form of a git repo) to ensure that I understand this exact issue correctly. My blind guess is that we don't recognize an inline `src` as "provided". Could you check if it works with this code?
```ts
/* parent-machine.ts */
import { createMachine, interpret } from 'xstate';
import { minuteMachine } from './minute-machine';

export const parentMachine = createMachine({
  id: 'parent',
  // eslint-disable-next-line @typescript-eslint/consistent-type-imports
  tsTypes: {} as import('./parent-machine.typegen').Typegen0,
  initial: 'pending',
  states: {
    pending: {
      invoke: {
        id: 'minuteMachine',
        src: 'minuteMachine',
        // The onDone transition will be taken when the
        // minuteMachine has reached its top-level final state.
        onDone: 'timesUp',
      },
    },
    timesUp: {
      type: 'final',
    },
  },
}, {
  services: {
    minuteMachine
  }
});

export const service = interpret(parentMachine);
```

**Reply by martynchamberlin** [reply]
Ah that made the error go away. Thanks @Andarist!

**Reply by Andarist** [reply]
It might still be worthwhile to report this [here](https://github.com/statelyai/xstate-tools)

**Reply by martynchamberlin** [reply]
Ok, fair enough. If my original syntax is expected to be supported, I'll see if I can put together a little reproducible repo and report it there. 👍
